prediction_market_agent/agents/microchain_agent/app.py

```diff
@@ -72,72 +76,110 @@ def maybe_initialize_agent(model: str) -> None:
+        model = st.selectbox(
+            "Model",
+            [st.session_state.agent.llm.generator.model],
+            index=0,
+            disabled=True,
+        )
+    model = check_not_none(model)
+
+    st.divider()
+    st.subheader("Built by Gnosis AI")
+    st.image(
+        "https://assets-global.website-files.com/63692bf32544bee8b1836ea6/63bea29e26bbff45e81efe4a_gnosis-owl-cream.png",
+        width=30,
     )
-if run_agent_button:
-    maybe_initialize_agent(model)
-    run_agent(
-        agent=st.session_state.agent,
-        iterations=int(iterations),
-        model=model,
+
+    st.caption(
+        "View the source code on our [github](https://github.com/gnosis/prediction-market-agent/tree/main/prediction_market_agent/agents/microchain_agent)"
     )
-if agent_is_initialized() and has_been_run_past_initialization(st.session_state.agent):
-    st.info(
-        "Run complete. Click 'Run' to allow the agent to continue, or add your "
-        "own reasoning."
+
+with st.expander(
+    "Interact with an autonomous agent that participates in prediction "
+    "markets. More info..."
+):
+    st.markdown(
+        "To start, click 'Run' to see the agent in action, or bootstrap the "
+        "agent with your own reasoning."
     )
-    # Display running cost
-    # st.info(f"Running OpenAPI credits cost: ${st.session_state.running_cost:.2f}")  # TODO debug why always == 0.0
-else:
-    st.info(
-        "Start by clicking 'Run' to see the agent in action. Alternatively "
-        "bootstrap the agent with your own reasoning before running."
+    st.markdown(
+        "It is equipped with the following tools to access the "
+        "[AIOmen](https://aiomen.eth.limo/) prediction market APIs:"
     )
+    st.markdown(get_market_function_bullet_point_list())
+
+# Placeholder for the agent's history
+history_container = st.container()
+
+# Interactive elements
+with bottom():
+    with st.container(border=True):
+        col1, col2, col3, col4 = st.columns([1, 1, 1, 8])
+        with col1:
+            run_agent_button = st.button("Run the agent")
+        with col2:
+            iterations = st.number_input(
+                "Iterations",
+                value=1,
+                step=1,
+                min_value=1,
+                max_value=100,
+            )
+        with col3:
+            st.caption(" \- OR -")
+        with col4:
+            user_reasoning = st.chat_input("Add reasoning...")
+
+# Execution
+with history_container:
+    if agent_is_initialized():
+        display_all_history(st.session_state.agent)
+    if user_reasoning:
+        maybe_initialize_agent(model)
+        execute_reasoning(
+            agent=st.session_state.agent,
+            reasoning=user_reasoning,
+            model=model,
+        )
+    if run_agent_button:
+        maybe_initialize_agent(model)
+        run_agent(
+            agent=st.session_state.agent,
+            iterations=int(iterations),
+            model=model,
```

Comment:
What's the difference between `run_agent` and `execute_reasoning`?

Reply:
- run_agent: The agent runs autonomously for N iterations
- execute_reasoning: The user of the app writes some reasoning to try and influence the steps of the agent (e.g. "Now see if I have a position in that market already, and if so sell it")

See this screenshot for how this maps to the UI

<img width="1431" alt="Screenshot 2024-05-08 at 15 39 28" src="https://github.com/gnosis/prediction-market-agent/assets/56087052/520586e9-1f73-4330-abd6-5416bb8b6caa">